### PR TITLE
[amdgpu] Add llvm.amdgcn.init.whole.wave intrinsic

### PR DESCRIPTION
**Pull request by rovka**
This intrinsic is meant to be used in functions that have a "tail" that
needs to be run with all the lanes enabled. The "tail" may contain
complex control flow that makes it unsuitable for the use of the
existing WWM intrinsics. Instead, we will pretend that the function
starts with all the lanes enabled, then branches into the actual body of
the function for the lanes that were meant to run it, and then finally
all the lanes will rejoin and run the tail.

As such, the intrinsic will return the EXEC mask for the body of the
function, and is meant to be used only as part of a very limited pattern
(for now only in amdgpu_cs_chain functions):

```
entry:
  %func_exec = call i1 @llvm.amdgcn.init.whole.wave()
  br i1 %func_exec, label %func, label %tail

func:
  ; ... stuff that should run with the actual EXEC mask
  br label %tail

tail:
  ; ... stuff that runs with all the lanes enabled;
  ; can contain more than one basic block
```

It's an error to use the result of this intrinsic for anything
other than a branch (but unfortunately checking that in the verifier is
non-trivial because SIAnnotateControlFlow will introduce an amdgcn.if
between the intrinsic and the branch).

The intrinsic is lowered to a SI_INIT_WHOLE_WAVE pseudo, which for now
is expanded in si-wqm (which is where SI_INIT_EXEC is handled too);
however the information that the function was conceptually started in
whole wave mode is stored in the machine function info
(hasInitWholeWave). This will be useful in prolog epilog insertion,
where we can skip saving the inactive lanes for CSRs (since if the
function started with all the lanes active, then there are no inactive
lanes to preserve).

### CODE REVIEW

**Comment by jayfoad** [comment]
Can this replacea llvm.amdgcn.init.exec? (Sorry if that's a dumb question, I have not grokked your new intrinsic yet.)

**Comment by rovka** [comment]
> Can this replacea llvm.amdgcn.init.exec? (Sorry if that's a dumb question, I have not grokked your new intrinsic yet.)

In theory, no, but I guess it depends on what use cases it actually has in practice. init.exec can take any exec mask, and I think it can exist anywhere in the function (the comment only says it gets moved to the beginning of the basic block, but it doesn't say it has to be the entry block). OTOH init.whole.wave should only appear in the entry block (maybe I could add a Verifier check for that if you think it's worth having) and enables all the lanes.

**Comment by Flakebi** [comment]
> Can this replacea llvm.amdgcn.init.exec? (Sorry if that's a dumb question, I have not grokked your new intrinsic yet.)

When used at the start of a function, their use-cases are rather different. `llvm.amdgcn.init.exec` **adds** an instruction that sets exec, `llvm.amdgcn.init.whole.wave` **removes** an instruction that sets exec (with the goal that the frontend can move the exec-setting to the end of the caller and the latency overlaps with the latency of the call).
Maybe there are cases where `llvm.amdgcn.init.exec` is currently used inside a function? (That could then be a candidate to use the new intrinsic, but it seems adventurous to use `llvm.amdgcn.init.exec` that way.)


Is there a test-case that does not use whole-wave-mode?
Due to whole-wave-mode, there’s a lot of exec-setting and prolog/epilog stuff going on.
I imagine the code that uses this new intrinsic does not use wwm as much, as the `tail` is effectively running in whole-wave-mode :)

```llvm
define amdgpu_cs_chain void @basic(<3 x i32> inreg %sgpr, ptr inreg %callee, i32 inreg %exec, { i32, ptr addrspace(5), i32, i32 } %vgpr, i32 %x, i32 %y) {
entry:
  %entry_exec = call i1 @llvm.amdgcn.init.whole.wave()
  br i1 %entry_exec, label %shader, label %tail

shader:
  %newx = add i32 %x, 42
  %oldval = extractvalue { i32, ptr addrspace(5), i32, i32 } %vgpr, 0
  %newval = add i32 %oldval, 5
  %newvgpr = insertvalue { i32, ptr addrspace(5), i32, i32 } %vgpr, i32 %newval, 0

  br label %tail

tail:
  %full.x = phi i32 [%x, %entry], [%newx, %shader]
  %full.vgpr = phi i32 [%vgpr, %entry], [%newvgpr, %shader]
  call void(ptr, i32, <3 x i32>, { i32, ptr addrspace(5), i32, i32 }, i32, ...) @llvm.amdgcn.cs.chain(ptr %callee, i32 %exec, <3 x i32> inreg %sgpr, { i32, ptr addrspace(5), i32, i32 } %full.vgpr, i32 %full.x)
  unreachable
}
```

File: llvm/include/llvm/IR/IntrinsicsAMDGPU.td [hunk 1]
```diff
@@ -208,6 +208,20 @@ def int_amdgcn_init_exec_from_input : Intrinsic<[],
...
+// resulting in a no-op. This pattern is used to optimize branches when function
+// tails need to be run in whole-wave-mode. It may also have other consequences
+// (mostly related to WWM CSR handling) that differentiate it from using
+// a plain `amdgcn.init.exec -1`.
+//
+// Can only be used in functions with the `amdgpu_cs_chain` calling convention.
+// Using this intrinsic without immediately branching on its return value is an
+// error.
+def int_amdgcn_init_whole_wave : Intrinsic<[llvm_i1_ty], [], [
+    IntrHasSideEffects, IntrNoMem, IntrNoDuplicate, IntrConvergent]>;
```
Comment (arsenm) [inline llvm/include/llvm/IR/IntrinsicsAMDGPU.td:223]: Does this really need IntrNoDuplicate, I would like to eliminate it

File: llvm/include/llvm/IR/IntrinsicsAMDGPU.td [hunk 2]
```diff
@@ -208,6 +208,20 @@ def int_amdgcn_init_exec_from_input : Intrinsic<[],
...
 
+// Sets the function into whole-wave-mode and returns whether the lane was
+// active when entering the function. A branch depending on this return will
+// revert the EXEC mask to what it was when entering the function, thus
+// resulting in a no-op. This pattern is used to optimize branches when function
+// tails need to be run in whole-wave-mode. It may also have other consequences
+// (mostly related to WWM CSR handling) that differentiate it from using
+// a plain `amdgcn.init.exec -1`.
+//
+// Can only be used in functions with the `amdgpu_cs_chain` calling convention.
```
Comment (arsenm) [inline llvm/include/llvm/IR/IntrinsicsAMDGPU.td:219]: The verifier needs to enforce this if it's a requirement

File: llvm/include/llvm/IR/IntrinsicsAMDGPU.td [hunk 3]
```diff
@@ -208,6 +208,20 @@ def int_amdgcn_init_exec_from_input : Intrinsic<[],
   [IntrConvergent, IntrHasSideEffects, IntrNoMem, IntrNoCallback,
    IntrNoFree, IntrWillReturn, ImmArg<ArgIndex<1>>]>;
 
+// Sets the function into whole-wave-mode and returns whether the lane was
+// active when entering the function. A branch depending on this return will
```
Comment (arsenm) [inline llvm/include/llvm/IR/IntrinsicsAMDGPU.td:212]: If we add the exec on entry to the function ABI (as we probably need to), can we remove the special calling convention requirement?

File: llvm/lib/Target/AMDGPU/SIISelLowering.cpp [hunk 4]
```diff
@@ -15671,6 +15671,133 @@ static int getAlignedAGPRClassID(unsigned UnalignedClassID) {
...
+        break;
+      }
+    }
+    assert(InactiveOpIdx != 0 && "Broken phi?");
+
+    // At this point, the register class could be larger than 32 or 64, so we
+    // might have to use more than one V_SET_INACTIVE instruction.
+    unsigned Size = TRI.getRegSizeInBits(*PhiRC);
+    switch (Size) {
+    case 32:
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/SIISelLowering.cpp:15727]: Braces

File: llvm/lib/Target/AMDGPU/SIISelLowering.cpp [hunk 5]
```diff
@@ -15671,6 +15671,133 @@ static int getAlignedAGPRClassID(unsigned UnalignedClassID) {
...
+      BuildMI(*EndBB, AfterEndCf, Phi.getDebugLoc(),
+              TII->get(AMDGPU::V_SET_INACTIVE_B64), PhiDest)
+          .addReg(NewPhiDest)
+          .add(Phi.getOperand(InactiveOpIdx));
+      break;
+    default: {
+      // For each 32-bit subregister of the register at InactiveOpIdx, insert
+      // a COPY to a new register, and a V_SET_INACTIVE_B32 using the
+      // corresponding subregisters of PhiDest and NewPhiDest.
+      // FIXME: There has to be a better way to iterate over this...
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/SIISelLowering.cpp:15743]: Is this just getSubRegFromChannel?

File: llvm/lib/Target/AMDGPU/SIInstructions.td [hunk 6]
```diff
@@ -575,6 +575,14 @@ def SI_INIT_EXEC_FROM_INPUT : SPseudoInstSI <
   let Defs = [EXEC];
 }
 
+// Sets EXEC to all lanes and returns the previous EXEC.
+def SI_INIT_WHOLE_WAVE : SPseudoInstSI <
+  (outs SReg_1:$dst), (ins),
+  [(set i1:$dst, (int_amdgcn_init_whole_wave))]> {
+  let Defs = [EXEC];
+  let Uses = [EXEC];
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/SIInstructions.td:583]: isConvergent = true?

File: llvm/include/llvm/IR/IntrinsicsAMDGPU.td (hunk 3 above)
Comment (rovka) [reply llvm/include/llvm/IR/IntrinsicsAMDGPU.td:212]: Sure, I don't think that's a hard requirement as much as an "I'm currently too lazy to think about/implement this for other calling conventions" requirement. I'm going to reword this.

**Comment by nhaehnle** [comment]
> > Can this replacea llvm.amdgcn.init.exec? (Sorry if that's a dumb question, I have not grokked your new intrinsic yet.)
> 
> In theory, no, but I guess it depends on what use cases it actually has in practice. init.exec can take any exec mask, and I think it can exist anywhere in the function (the comment only says it gets moved to the beginning of the basic block, but it doesn't say it has to be the entry block). OTOH init.whole.wave should only appear in the entry block (maybe I could add a Verifier check for that if you think it's worth having) and enables all the lanes.

I think it could replace `init.exec`. init.exec is also supposed to appear in the entry block,and the statement about it being moved up is really a requirement for this new intrinsic as well -- we want to be robust against intermediate transforms adding something above the intrinsic, perhaps because they added an `alloca` or something.

**Review by nhaehnle — COMMENTED** [review]
Thanks! We've just had a quick offline discussion. To summarize, I think the overall flow of where stuff happens is good, but the optimization of the initial branch is subtle. I suggest splitting this up into multiple changes, where the first one implements init.whole.wave in a simple but hopefully obviously correct manner. The optimization of the initial branch can then be done as a separate change.

I also think we should try to find an alternative to using V_SET_INACTIVE, since the whole WWM infrastructure is awfully fragile. The closer we can get to "normal looking" IR between isel and register allocation, the better.

**Comment by ruiling** [comment]
I don't think we can simply remove the init.exec.wave in the entry block. Think about the case like below where we have a use of the function argument in tail block:
```
define amdgpu_cs_chain void @basic(<3 x i32> inreg %sgpr, ptr inreg %callee, i32 inreg %exec, { i32, ptr addrspace(5), i32, i32 } %vgpr) {
entry:
  %entry_exec = call i1 @llvm.amdgcn.init.whole.wave()
  br i1 %entry_exec, label %shader, label %tail

shader:
   ...
  br label %tail

tail:
  %x = extractelement <16 x i32> %vgpr, i32 0
  %y = add i32 %x, 32    ;<-- the instruction should operate on all the lanes of %x
  %vgpr.new = insertelement <16 x i32> %vgpr, i32 %y, i32 0
  call void(ptr, i32, <3 x i32>, { i32, ptr addrspace(5), i32, i32 }, i32, ...) @llvm.amdgcn.cs.chain(ptr %callee, i32 %exec, <3 x i32> inreg %sgpr, { i32, ptr addrspace(5), i32, i32 } %vgpr.new)
  unreachable
}
```
This would be very likely lowered to something like: (Sorry it is in very rough shape.)
```
 entry:
   ; setup exec for whole wave mode
   %arg = COPY $vgpr8
   br %old_exec, %shader, %tail

shader:
 ....

tail:
 %new_arg = V_ADD_I32 %arg, 32
 SI_CS_CHAIN .... %new_arg
```
The point is when have an instruction in tail block which want to operate on a function argument, I think the expectation is the `V_ADD_I32` here should operate(read/write) all the lanes. If we remove the exec setup in the entry block, the instruction would not be able to see the values in the lanes that were inactive at function start. Does this make sense?

File: llvm/include/llvm/IR/IntrinsicsAMDGPU.td [hunk 7]
```diff
@@ -208,6 +208,19 @@ def int_amdgcn_init_exec_from_input : Intrinsic<[],
...
+// revert the EXEC mask to what it was when entering the function, thus
+// resulting in a no-op. This pattern is used to optimize branches when function
+// tails need to be run in whole-wave-mode. It may also have other consequences
+// (mostly related to WWM CSR handling) that differentiate it from using
+// a plain `amdgcn.init.exec -1`.
+//
+// Using this intrinsic without immediately branching on its return value is an
+// error.
```
Comment (nhaehnle) [inline llvm/include/llvm/IR/IntrinsicsAMDGPU.td:220]: Remove this part of the comment.

File: llvm/lib/Target/AMDGPU/SIWholeQuadMode.cpp [hunk 8]
```diff
@@ -1582,6 +1583,28 @@ void SIWholeQuadMode::lowerInitExec(MachineInstr &MI) {
   MachineBasicBlock *MBB = MI.getParent();
   bool IsWave32 = ST->isWave32();
 
+  if (MI.getOpcode() == AMDGPU::SI_INIT_WHOLE_WAVE) {
+    // TODO: Assert that it's in the entry block
```
Comment (nhaehnle) [inline llvm/lib/Target/AMDGPU/SIWholeQuadMode.cpp:1587]: Do the TODO? :)

File: llvm/test/CodeGen/AMDGPU/llvm.amdgcn.init.whole.wave-w32.ll [hunk 9]
```diff
@@ -0,0 +1,1127 @@
...
+; GISEL12-NEXT:    s_mov_b32 s6, s3
+; GISEL12-NEXT:    s_mov_b32 s7, s4
+; GISEL12-NEXT:    s_wait_alu 0xfffe
+; GISEL12-NEXT:    s_and_saveexec_b32 s3, s8
+; GISEL12-NEXT:  ; %bb.1: ; %shader
+; GISEL12-NEXT:    v_add_nc_u32_e32 v12, 42, v12
+; GISEL12-NEXT:    v_add_nc_u32_e32 v8, 5, v8
+; GISEL12-NEXT:  ; %bb.2: ; %tail
+; GISEL12-NEXT:    s_wait_alu 0xfffe
+; GISEL12-NEXT:    s_or_b32 exec_lo, exec_lo, s3
```
Comment (nhaehnle) [inline llvm/test/CodeGen/AMDGPU/llvm.amdgcn.init.whole.wave-w32.ll:25]: We could just `s_mov_b32 exec_lo, -1` here. Not particularly high priority, and if anything should happen as part of a more general micro-optimization that detects the equivalent condition in compute shaders with known workgroup size.

**Review by nhaehnle — COMMENTED** [review]
Thanks! Two nits inline, but apart from that LGTM.

File: llvm/lib/Target/AMDGPU/SIWholeQuadMode.cpp (hunk 8 above)
Comment (rovka) [reply llvm/lib/Target/AMDGPU/SIWholeQuadMode.cpp:1587]: Ooops, I thought for sure I would notice that before pushing...

**Review by nhaehnle — APPROVED** [review]
Thanks, LGTM!

**Pull request merged**